File: 010.regular-expression-matching/regular-expression-matching.js
/**
 * @param {string} s
 * @param {string} p
 * @return {boolean}
 */
// memo 自顶而下
var isMatch = function (s, p) {
  let memo = {}
  const match = (i, j) => {
    if (`${i},${j}` in memo) {
      return memo[`${i},${j}`]
    }
    if (j === p.length) {
      return i === s.length
    }
    const firstMatch = i < s.length && (p[j] == '.' || p[j] == s[i])
    if (j + 1 < p.length && p[j + 1] == '*') {
      memo[`${i},${j}`] = match(i, j + 2) || firstMatch && match(i + 1, j)
      return memo[`${i},${j}`]
    }
    if (firstMatch) {
      memo[`${i},${j}`] = match(i + 1, j + 1)
      return memo[`${i},${j}`]
    }
    memo[`${i},${j}`] = false
    return memo[`${i},${j}`]
  }
  return match(0, 0)
};

console.log(isMatch("aa", "a") == false)
console.log(isMatch("aa", "a*") == true)
console.log(isMatch("ab", ".*") == true)
console.log(isMatch("fghijkae", ".*ae") == true)
console.log(isMatch("aab", "c*a*b") == true)
console.log(isMatch("mississippi", "mis*is*p*.") == false)
console.log(isMatch("aaa", "a*a") == true)
console.log(isMatch("aaa", "ab*a*c*a") == true)
console.log(isMatch("abbabaaaaaaacaa", "a*.*b.a.*c*b*a*c*") == true)
console.log(isMatch("bcbabcaacacbcabac", "a*c*a*b*.*aa*c*a*a*") == true)

File: hard/010.regular-expression-matching/regular-expression-matching.py
# 解法1：根据第一个位置情况分类
class Solution:

    def isMatch(self, s: str, p: str) -> bool:
        if not (bool(s) | bool(p)):
            return True
        if s and not p:
            return False
        if not s and p:
            if len(p) > 1 and p[1] == '*':
                return self.isMatch(s, p[2:])
            return False

        if p[0] == s[0]:
            if len(p) > 1 and p[1] == '*':
                if len(p) > 2:
                    i = 0
                    l = len(s)
                    while i < l:
                        if self.isMatch(s[i:], p[2:]):
                            return True
                        if s[i] != p[0]:
                            return False
                        i += 1
                    if len(p) % 2 == 1 or not all(map(lambda x: x == '*', p[1::2])):
                        return False
                for i in s:
                    if i != p[0]:
                        return False
                return True
            return self.isMatch(s[1:], p[1:])

        if p[0] == '.':
            if len(p) > 1 and p[1] == '*':
                if len(p) > 2:
                    i = 0
                    l = len(s)
                    while i < l:
                        if self.isMatch(s[i:], p[2:]):
                            return True
                        i += 1
                    if len(p) % 2 == 0 and all(map(lambda x: x == '*', p[1::2])):
                        return True
                    return False
                return True
            return self.isMatch(s[1:], p[1:])

        if len(p) > 1 and p[1] == '*':
            return self.isMatch(s, p[2:])
        return False


if __name__ == '__main__':
    s = Solution()
    isMatch = s.isMatch

    print(isMatch("aa", "a") == False)
    print(isMatch("aa", "a*") == True)
    print(isMatch("ab", ".*") == True)
    print(isMatch("fghijkae", ".*ae") == True)
    print(isMatch("aab", "c*a*b") == True)
    print(isMatch("mississippi", "mis*is*p*.") == False)
    print(isMatch("aaa", "a*a") == True)
    print(isMatch("aaa", "ab*a*c*a") == True)
    print(isMatch("abbabaaaaaaacaa", "a*.*b.a.*c*b*a*c*") == True)
    print(isMatch("bcbabcaacacbcabac", "a*c*a*b*.*aa*c*a*a*") == True)
